### PR TITLE
[new-exe] convert fused_all_reduce_op_handle to program

### PR DESCRIPTION
**Pull request by zhiqiu**
### PR types

New features
### PR changes

Others
### Describe

Support convertting `fused_all_reduce_op_handle` in Graph to `fake_coalesce + c_allreduce_sum` in Program.

### CODE REVIEW

**Review by Aurelius84 — DISMISSED** [review]

File: paddle/phi/kernels/fake_coalesce_kernel.cc [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+template <typename T, typename Context>
+void FakeCoalesceKernel(const Context &dev_ctx,
+                        const std::vector<const DenseTensor *> &input,
+                        DenseTensor *output,
+                        std::vector<DenseTensor *> xout) {
+  int64_t size_of_dtype = sizeof(T);
+  auto dtype = input.at(0)->dtype();
+  int64_t numel = 0;
+  // check address
+  for (size_t i = 1; i < input.size(); ++i) {
```
Comment (From00) [inline paddle/phi/kernels/fake_coalesce_kernel.cc]: I see a sort before checking address in fused_all_reduce_op_handle, is it no longer needed here?

File: paddle/phi/kernels/fake_coalesce_kernel.cc [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/phi/kernels/fake_coalesce_kernel.h"
+
+#include <sstream>
+#include <vector>
+#include "glog/logging.h"
+
+#include "paddle/phi/backends/cpu/cpu_context.h"
```
Comment (chenwhql) [inline paddle/phi/kernels/fake_coalesce_kernel.cc]: 这两个context可以不include，kernel_registry.h有全部的

File: paddle/phi/kernels/fake_coalesce_kernel.cc [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+#include "paddle/phi/backends/gpu/gpu_context.h"
+#include "paddle/phi/core/kernel_registry.h"
+
+#include "paddle/fluid/platform/device_memory_aligment.h"
+
+namespace phi {
+
+template <typename T, typename Context>
+void FakeCoalesceKernel(const Context &dev_ctx,
```
Comment (chenwhql) [inline paddle/phi/kernels/fake_coalesce_kernel.cc]: 已有coalesce_tensor_kernel了，可以复用吗

File: paddle/phi/kernels/fake_coalesce_kernel.cc [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            "The infered address of the next tensor should be equal to the "
+            "real address of the next tensor. But got infered address is %p "
+            "and real address is %p.",
+            infer_next_address,
+            next_address));
+  }
+  numel += paddle::platform::Alignment(
+      (*input.rbegin())->numel() * size_of_dtype, dev_ctx.GetPlace());
+  // reset holder, do inplace
+  output->ShareBufferWith(*input.at(0));
```
Comment (From00) [inline paddle/phi/kernels/fake_coalesce_kernel.cc]: It seems that fake_colesce OP is in-place implicitly, should its input and output vars be considered to skip in-place in interpretercore?

File: paddle/fluid/framework/ir/graph_helper.cc [hunk 5]
```diff
@@ -497,6 +505,70 @@ static OpDesc *ReplaceScaleLossGradOp(const Node &node, OpDesc *desc) {
   return desc;
 }
 
+static void ReplaceAllReduceOp(const Node &node,
+                               proto::BlockDesc *block,
+                               std::vector<OpDesc> *ops) {
+#if defined(PADDLE_WITH_NCCL) || defined(PADDLE_WITH_RCCL)
+  ops->emplace_back();
+  auto &desc1 = ops->back();
+  std::string name = "fake_coalesce_" + std::to_string(ops->size());
```
Comment (From00) [inline paddle/fluid/framework/ir/graph_helper.cc:514]: Not important, the _name_ is not used until line 537 in _if_ statement, better to be constructed inside the _if_ statement.

File: paddle/fluid/framework/ir/graph_helper.cc [hunk 6]
```diff
@@ -497,6 +505,70 @@ static OpDesc *ReplaceScaleLossGradOp(const Node &node, OpDesc *desc) {
...
+    desc2.SetAttr("use_calc_stream", true);
+  }
+
+  desc1.SetAttr(OpProtoAndCheckerMaker::OpRoleAttrName(),
+                (static_cast<int>(OpRole::kBackward)));
+  desc2.SetAttr(OpProtoAndCheckerMaker::OpRoleAttrName(),
+                (static_cast<int>(OpRole::kBackward)));
+#else
+  PADDLE_THROW(platform::errors::Unimplemented(
+      "ReplaceAllReduceOp is only implemented for paddle compiled with GPU."));
```
Comment (From00) [inline paddle/fluid/framework/ir/graph_helper.cc]: Not important, "compiled with GPU" should be "compiled with NCCL".

File: paddle/phi/kernels/fake_coalesce_kernel.cc (hunk 1 above)
Comment (zhiqiu) [reply paddle/phi/kernels/fake_coalesce_kernel.cc]: To my knowledge，the input tensor is already ordered.

File: paddle/phi/kernels/fake_coalesce_kernel.cc (hunk 4 above)
Comment (zhiqiu) [reply paddle/phi/kernels/fake_coalesce_kernel.cc]: Basically, Yes! However, in the specific case, the inputs and outputs are grad of parameter, we do not to handdle it and the `coalesce_tensor` and `c_allreduce_sum` are in-place implicitly as well as `fake_coalesce`.

**Review by From00 — DISMISSED** [review]
LGTM

File: paddle/phi/api/yaml/api.yaml [hunk 7]
```diff
@@ -108,6 +108,14 @@
   inplace : (x -> out)
   backward : erfinv_grad
 
+- api : fake_coalesce
```
Comment (chenwhql) [inline paddle/phi/api/yaml/api.yaml]: 我们能否给它一个准确一些的命名，比如memory_contiguous之类的？fake_coalesce这个名字看起来属于trick的接口，暂时不建议加到api.yaml里面，如果一定需要，可以仍然使用OpMaker去新增

然后，建议加个注释说明下这个不会序列化到推理program中，api.yaml中还是希望尽可能是python->C++一致的接口声明，但确实这种功能性的算子也是切实存在的需求，后续我们再讨论下是否要将它直接放到api.yaml中，可能会移走

Comment (zhiqiu) [reply paddle/phi/api/yaml/api.yaml]: 嗯，后面可以搞一个 op.yaml 之类的，把它迁过去。

File: paddle/phi/kernels/fake_coalesce_kernel.cc (hunk 3 above)
Comment (zhiqiu) [reply paddle/phi/kernels/fake_coalesce_kernel.cc]: 逻辑不太一样。

File: paddle/phi/kernels/fake_coalesce_kernel.cc (hunk 2 above)
Comment (zhiqiu) [reply paddle/phi/kernels/fake_coalesce_kernel.cc]: done

File: paddle/fluid/framework/ir/graph_helper.cc (hunk 6 above)
Comment (zhiqiu) [reply paddle/fluid/framework/ir/graph_helper.cc]: done

File: paddle/fluid/framework/ir/graph_helper.cc (hunk 5 above)
Comment (zhiqiu) [reply paddle/fluid/framework/ir/graph_helper.cc:514]: It is for the correct ops->size().

**Review by chenwhql — APPROVED** [review]

**Review by From00 — APPROVED** [review]

**Pull request merged**